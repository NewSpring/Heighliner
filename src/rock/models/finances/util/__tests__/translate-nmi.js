import { defaultsDeep } from "lodash";

import translate, { getCardType } from "../translate-nmi";

import {
  standard,
  recurring,
  check,
  NMIExample,
} from "../__mocks__/sample-response";

jest.mock("moment", () => date => ({
  toISOString: () => `Mocked ISODate: ${date}`,
}));

describe("getCardType", () => {
  it("correctly identifies a visa with hashes", () => {
    expect(getCardType("4111********1111")).toBe(7);
  });

  it("correctly identifies a visa", () => {
    const cards = ["4111111111111111", "4012888888881881", "4222222222222"];
    cards.map(x => expect(getCardType(x)).toBe(7));
  });

  it("correctly identifies a mastercard", () => {
    const cards = ["5555555555554444", "5105105105105100", "5431111111111111"];
    cards.map(x => expect(getCardType(x)).toBe(8));
  });

  it("correctly identifies an American Express", () => {
    const cards = ["378282246310005", "371449635398431", "341111111111111"];
    cards.map(x => expect(getCardType(x)).toBe(159));
  });

  it("correctly identifieds a discover card", () => {
    const cards = ["6011111111111117", "6011000990139424", "6011601160116611"];
    cards.map(x => expect(getCardType(x)).toBe(160));
  });

  it("returns a falsy if no valid card is found", () => {
    expect(getCardType("xxxxx")).toBeFalsy();
  });
});

jest.mock("node-uuid", () => ({
  v4: () => "a31044f3-d721-47b2-a91d-e58ac41832ad",
}));

describe("translate", () => {
  it("correctly formats a standard transaction", () => {
    expect(translate(standard, { Id: 3 })).toMatchSnapshot();
  });

  it("correctly formats the NMI example transaction", () => {
    expect(translate(NMIExample, { Id: 3 })).toMatchSnapshot();
  });

  it("returns null for a pending transaction", () => {
    const example = defaultsDeep({ condition: "pending" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a pendingsettlement transaction", () => {
    const example = defaultsDeep({ condition: "pendingsettlement" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for an in_progress transaction", () => {
    const example = defaultsDeep({ condition: "in_progress" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for an abandoned transaction", () => {
    const example = defaultsDeep({ condition: "abandoned" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a failed transaction", () => {
    const example = defaultsDeep({ condition: "failed" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  xit("returns correctly for a failed transaction", () => {
    const example = defaultsDeep({ condition: "failed" }, standard);
    expect(translate(example, { Id: 3 })).toMatchSnapshot();
  });

  it("returns null for a canceled transaction", () => {
    const example = defaultsDeep({ condition: "canceled" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for an unknown transaction", () => {
    const example = defaultsDeep({ condition: "unknown" }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a refund transaction", () => {
    const example = defaultsDeep(
      { action: { action_type: "refund" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  xit("correctly formats refunds", () => {
    const example = defaultsDeep(
      { action: { action_type: "refund" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toMatchSnapshot();
  });

  it("returns null for a credit transaction", () => {
    const example = defaultsDeep(
      { action: { action_type: "credit" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a auth transaction", () => {
    const example = defaultsDeep({ action: { action_type: "auth" } }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a capture transaction", () => {
    const example = defaultsDeep(
      { action: { action_type: "capture" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a void transaction", () => {
    const example = defaultsDeep({ action: { action_type: "void" } }, standard);
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  it("returns null for a return transaction", () => {
    const example = defaultsDeep(
      { action: { action_type: "return" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toBeFalsy();
  });

  xit("correctly formats a return action", () => {
    const example = defaultsDeep(
      { action: { action_type: "return" } },
      standard
    );
    expect(translate(example, { Id: 3 })).toMatchSnapshot();
  });

  it("allows sale to not be the first action", () => {
    const example = defaultsDeep({}, NMIExample);
    const sale = example.action[0];
    const other = example.action[1];
    const settle = example.action[2];
    example.action = [other, settle, sale];
    expect(translate(example, { Id: 3 })).toMatchSnapshot();
  });

  it("supports multiple funds", () => {
    const example = defaultsDeep(
      {
        product: [
          standard.product,
          {
            sku: "128",
            quantity: "10.0000",
            description: "Step Up",
          },
        ],
      },
      standard
    );
    expect(translate(example, { Id: 3 })).toMatchSnapshot();
  });

  it("supports a recurring transaction", () => {
    expect(translate(recurring, { Id: 3 })).toMatchSnapshot();
  });

  it("supports manually setting a person", () => {
    expect(translate(recurring, { Id: 3 }, 10)).toMatchSnapshot();
  });

  it("supports a check", () => {
    expect(translate(check, { Id: 3 })).toMatchSnapshot();
  });
});
