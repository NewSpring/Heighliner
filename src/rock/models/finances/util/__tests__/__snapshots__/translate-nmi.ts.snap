exports[`translate allows sale to not be the first action 1`] = `
Object {
  "CampusId": undefined,
  "Location": Object {
    "City": "New York City",
    "Country": "US",
    "PostalCode": "10001",
    "State": "NY",
    "Street1": "123 Main St",
    "Street2": "Apt B",
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "johnsmith@example.com",
    "LastName": "Smith",
  },
  "ScheduledTransaction": Object {},
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "1234567890",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 2612675976",
    "TransactionCode": "2612675976",
    "TransactionDateTime": "2015-03-13T01:52:05.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": NaN,
      "Amount": 11,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate correctly formats a standard transaction 1`] = `
Object {
  "CampusId": 16,
  "Location": Object {
    "City": "City",
    "Country": undefined,
    "PostalCode": "29621",
    "State": "SC",
    "Street1": "Address",
    "Street2": undefined,
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "test@email.com",
    "LastName": "Last",
  },
  "ScheduledTransaction": Object {},
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "apollos_1",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 1",
    "TransactionCode": "1",
    "TransactionDateTime": "2016-11-02T13:08:45.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": 125,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate correctly formats the NMI example transaction 1`] = `
Object {
  "CampusId": undefined,
  "Location": Object {
    "City": "New York City",
    "Country": "US",
    "PostalCode": "10001",
    "State": "NY",
    "Street1": "123 Main St",
    "Street2": "Apt B",
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "johnsmith@example.com",
    "LastName": "Smith",
  },
  "ScheduledTransaction": Object {},
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "1234567890",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 2612675976",
    "TransactionCode": "2612675976",
    "TransactionDateTime": "2015-03-13T01:52:05.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": NaN,
      "Amount": 11,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate supports a check 1`] = `
Object {
  "CampusId": 16,
  "Location": Object {
    "City": "City",
    "Country": undefined,
    "PostalCode": "29621",
    "State": "SC",
    "Street1": "Address",
    "Street2": undefined,
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "1****1111",
    "CurrencyTypeValueId": 157,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "test@email.com",
    "LastName": "Last",
  },
  "ScheduledTransaction": Object {},
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "apollos_1",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 1",
    "TransactionCode": "1",
    "TransactionDateTime": "2016-11-02T13:08:45.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": 125,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate supports a recurring transaction 1`] = `
Object {
  "CampusId": 16,
  "Location": Object {
    "City": "City",
    "Country": undefined,
    "PostalCode": "29621",
    "State": "SC",
    "Street1": "Address",
    "Street2": undefined,
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "test@email.com",
    "LastName": "Last",
  },
  "ScheduledTransaction": Object {
    "GatewayScheduleId": "10",
  },
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "apollos_1",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 1",
    "TransactionCode": "1",
    "TransactionDateTime": "2016-11-02T13:08:45.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": 125,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate supports manually setting a person 1`] = `
Object {
  "CampusId": 16,
  "Location": Object {
    "City": "City",
    "Country": undefined,
    "PostalCode": "29621",
    "State": "SC",
    "Street1": "Address",
    "Street2": undefined,
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Id": 10,
  },
  "ScheduledTransaction": Object {
    "GatewayScheduleId": "10",
  },
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "apollos_1",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 1",
    "TransactionCode": "1",
    "TransactionDateTime": "2016-11-02T13:08:45.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": 125,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;

exports[`translate supports multiple funds 1`] = `
Object {
  "CampusId": 16,
  "Location": Object {
    "City": "City",
    "Country": undefined,
    "PostalCode": "29621",
    "State": "SC",
    "Street1": "Address",
    "Street2": undefined,
  },
  "PaymentDetail": Object {
    "AccountNumberMasked": "4***********1111",
    "CreditCardTypeValueId": 7,
    "CurrencyTypeValueId": 156,
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
  },
  "Person": Object {
    "Email": "test@email.com",
    "LastName": "Last",
  },
  "ScheduledTransaction": Object {},
  "Transaction": Object {
    "FinancialGatewayId": 3,
    "ForeignKey": "apollos_1",
    "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    "SourceTypeValueId": 798,
    "Summary": "Reference Number: 1",
    "TransactionCode": "1",
    "TransactionDateTime": "2016-11-02T13:08:45.000Z",
    "TransactionTypeValueId": 53,
  },
  "TransactionDetails": Array [
    Object {
      "AccountId": 125,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
    Object {
      "AccountId": 128,
      "Amount": 1,
      "Guid": "a31044f3-d721-47b2-a91d-e58ac41832ad",
    },
  ],
}
`;
